agents/craftassist/make_swarm.py

```diff
@@ -0,0 +1,367 @@
+            try:
+                if all(self.init_status):
+                    self.base_agent.step()
+                self.handle_worker_perception()
+                self.assign_new_tasks_to_workers()
+                self.update_tasks_with_worker_data()
+                self.handle_worker_memory_queries()
+                                
+            except Exception as e:
+                self.base_agent.handle_exception(e)
+
+
+class SwarmWorkerWrapper(Process):
+    def __init__(self, opts, task_map, disable_perception_modules, idx=0) -> None:
+        super().__init__()
+        self.opts = opts
+        self.idx = idx
+        self.input_tasks = Queue()
+        self.perceptions = Queue()
+        self.query_from_worker = Queue()
+        self.query_from_master = Queue()
+        self.memory_send_queue = Queue()
+        self.memory_receive_queue = Queue()
+        self.init_task_map(task_map)
+        self.disable_perception_modules = disable_perception_modules
+        
+    def init_task_map(self, task_map, task_info=None):
+        self.TASK_MAP = deepcopy(task_map)
+        self.TASK_INFO = get_default_task_info(task_map)
+        if task_info is not None:
+            for key in task_info:
+                self.TASK_INFO[key] = task_info[key]
+
+    def init_worker(self, agent):
+        self.agent_type = agent.__class__.__name__.lower()
+
+        agent.agent_idx = self.idx
+        agent.task_stacks = dict()
+        agent.task_ghosts = []
+        agent.prio = dict()
+        agent.running = dict()
+        agent.pause = dict()
+        agent.memory_send_queue = self.memory_send_queue
+        agent.memory_receive_queue = self.memory_receive_queue
+        agent.query_from_worker = self.query_from_worker
+
+        # # craftassist 
+        # p = agent.get_player()
+        # agent.entityId = p.entityId
+
+        # disable perception modules
+        for module_key in self.disable_perception_modules:
+            del agent.perception_modules[module_key]
+        
+        #### temporary for debug
+        agent.perception_modules = dict()
+        agent.perception_modules["low_level"] = SwarmLowLevelMCPerception(agent)
+        #### end temporary for debug
+        
+        # memory
+        agent.memory = SwarmWorkerMemory(memory_send_queue=self.memory_send_queue,
+                                         memory_receive_queue=self.memory_receive_queue,
+                                         memory_tag="swarm_worker_{}".format(agent.agent_idx))        
+        # controller
+        agent.disable_chat = True
+    
+    def check_task_info(self, task_name, task_data):
+        if task_name not in self.TASK_INFO.keys():
+            logging.info("task {} received without checking arguments")
+            return True
+        for key in self.TASK_INFO[task_name.lower()]:
+            if key not in task_data:
+                return False
+        return True
+
+    def preprocess_data(self, task_name, task_data):
```

Comment:
not sure I understand what this does

Reply:
Just preprocessing task data. Sometimes information is already in the input argument task_data, sometimes the information is in task_data["task_data"]. Different tasks have different formats.